### ISSUE TITLE
Add doxygen comments for the purpose to as many functions as possible

### ISSUE THREAD

**Issue by toogley** [neomutt/neomutt#199]
One of our long-term goals is to automatically generate documentation about what every function in our code does.

For that, i think it makes sense to add to every function we know the behavior like e.g. javadoc comments. This makes it easier for new developers to understand the code, e.g. as they don't have to guess what a function/argument does. 

The exact format of the docstring doesn't matter, as long its consistent. Later on, when we are more sure about how to generate the code-documentation, we can change the format easily.

To prove that this currently happens not every time: [This commit](https://github.com/neomutt/neomutt/commit/512a6d4286af802f07b5a82067ca2018de7f3c3a#diff-f7290c71af5bad3f7cda9d9253e9ecb4R1097) introduced two new functions `mutt_is_inbox` and `mutt_same_path` which were not really documented. Just to make my intension more clear: @gahr @flatcap i don't want to discourage/negatively influence both of you, i just want to improve our great project :)
- **What do you think about that? Does it really add value to our Source Code?**
-  **Format**

I propose a more structured Format, like [javadoc](https://en.wikipedia.org/wiki/Javadoc#Example):

``` java
/**
 * Validates a chess move.
 *
 * Use {@link #doMove(int theFromFile, int theFromRank, int theToFile, int theToRank)} to move a piece.
 *
 * @param theFromFile file from which a piece is being moved
 * @param theFromRank rank from which a piece is being moved
 * @param theToFile   file to which a piece is being moved
 * @param theToRank   rank to which a piece is being moved
 * @return            true if the move is valid, otherwise false
 */
boolean isValidMove(int theFromFile, int theFromRank, int theToFile, int theToRank) {
```

Of course it isn't really useful, if we just repeat the information which is already given by the names of the function and the variables and the types - instead, I estimtate this as most useful if we really describe what this variable really does, ideally using differnt words than those in the variable names.
- **What was meant [here with "Public function" ](https://github.com/neomutt/neomutt/blob/neomutt/browser.c#L890) ?**

**Comment by flatcap** [comment]
I've been a bit lazy, in this respect.

Re-writing the Sidebar, I carefully structured the code and [documented all the functions](https://github.com/neomutt/neomutt/blob/a1613a0/sidebar.c#L274-L283).
Many of the features got the same treatment.

The style I've used is parsed by [Doxygen](http://www.stack.nl/~dimitri/doxygen/).  (I like the diagrams it produces :-)

However, I haven't been so rigorous with newer code.
I shall insist in future that any function/struct we introduce must be documented.

We should also create a list of all the undocumented functions that we've introduced and prioritise them for documenting.

**Comment by pickfire** [comment]
@toogley I will still go with Doxygen even though I doesn't know how to use it but then it's build for c projects.

**Comment by gahr** [comment]
Also, should non-static functions be documented at point of declaration or definition?

**Comment by flatcap** [comment]
I've always documented at the point of **definition**.  Then a tool creates pretty docs for the developer.

But, I can see a good argument for the headers being documented -- that one file is the API & docs.

**Comment by flatcap** [comment]
There are ~2000 functions in Mutt.  NeoMutt adds a further 255.
The Sidebar and Compress features are well documented, the rest less so.

Through the power of scripting, I've added template comment blocks to all the functions.
Yes, all 2000 of them.  There will be quite a few duplicate blocks, for now.
The templates include the function name, arguments and return type.
- 133 files changed, 17549 insertions

**Comment by flatcap** [comment]
Here's a list of some of commands I used to generate the data for the function blocks.
(then I used a LOT of vim macros to put things in place).
```
ctags --list-kinds=c

find mutt neo -type f ! -name '*.[ch]' -delete
find . -type l -delete

# Standalone programs
dotlock.c md5.c pgpewrap.c pgppubring.c txt2c.c doc/makedoc.c

# Functions
cd mutt; ctags -R -x --c-kinds=f . > ../m-fun1; cd ..
cd neo;  ctags -R -x --c-kinds=f . > ../n-fun1; cd ..

# Enumerations, structs, unions
cd mutt; ctags -R -x --c-kinds=gsu . > ../m-struct1; cd ..
cd neo;  ctags -R -x --c-kinds=gsu . > ../n-struct1; cd ..

# Variables
cd mutt; ctags -R -x --c-kinds=v . > ../m-var1; cd ..
cd neo;  ctags -R -x --c-kinds=v . > ../n-var1; cd ..

for i in *.c imap/*.c; do
	cproto -D HAVE_CONFIG_H=1 -I . -I imap -s $i
	echo
done > p1.txt
```

**Comment by flatcap** [comment]
I've filtered the undocumented list down to just the stuff we've introduced: 174 functions, 389 variables - [devel/doxygen-neo](https://github.com/neomutt/neomutt/tree/devel/doxygen-neo)
Any suggestions how we tackle this?

**Comment by nuclearsandwich** [comment]
> Any suggestions how we tackle this?

I was giving this some thought during my first contributions. Do we have any developer docs beyond <https://github.com/neomutt/neomutt/blob/neomutt/doc/devel-notes.txt>?
Making sure that it's straightforward and encouraged for new contributors like myself to document our additions by including how to do so in developer docs will go a long way to make sure new stuff doesn't add to the problem.

As far as paying down the doc-comment deficit, as I read through and touch various new-to-me parts of the code, I can try to document. If multiple contributors  make sure that functions they touch have documentation then we can close the gap more naturally and getting docs on areas of the code that are most active and won't feel as [Sisyphean](https://en.wikipedia.org/wiki/Sisyphus) as trying to walk through and just document everything.

**Comment by flatcap** [comment]
> Do we have any developer docs beyond devel-notes.txt

There are some scattered notes, but not really.
That file's inherited from Mutt and could really do with an overhaul.

> Making sure that it's straightforward and encouraged for new contributors like myself to document our additions...

Absolutely.  I created templates for documenting the features in the [management](https://github.com/neomutt/management) repo.
I can create a sample C function header.

> doc-comment deficit

A year ago, I offered to fix upstream Mutt's whitespace.  They wouldn't accept it.
They decided that they're rather let it get better incrementally.   This will never happen.

For **our** code I'm tempted to put on some crap movies and document all the functions myself.
Just get it done.

> contributors make sure that functions they touch have documentation ... getting docs on areas of the code that are most active

That's a good idea.

Also, we should prioritise the commonly used `struct`s to make it easier to understand the bigger picture.

> and won't feel as **Sisyphean** as trying to ... document everything.

Good classical education reference :-)
